pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,327 @@
+	ForceSingleItem bool // Force the use of the single item iterator even if the batch API is supported
+}
+
+// Put inserts a new key/value pair in the map. If the key already exists, the value is updated
+func (g *GenericMap[K, V]) Put(key *K, value *V) error {
+	if g.isPerCPU() {
+		return g.m.Put(unsafe.Pointer(key), *value)
+	}
+
+	return g.m.Put(unsafe.Pointer(key), unsafe.Pointer(value))
+}
+
+// Update updates the value of an existing key in the map.
+func (g *GenericMap[K, V]) Update(key *K, value *V, flags ebpf.MapUpdateFlags) error {
+	return g.m.Update(unsafe.Pointer(key), unsafe.Pointer(value), flags)
+}
+
+// Lookup looks up a key in the map and returns the value. If the key doesn't exist, it returns ErrKeyNotExist
+func (g *GenericMap[K, V]) Lookup(key *K, valueOut *V) error {
+	if g.isPerCPU() {
+		return g.m.Lookup(unsafe.Pointer(key), *valueOut)
+	}
+
+	return g.m.Lookup(unsafe.Pointer(key), unsafe.Pointer(valueOut))
+}
+
+// Delete deletes a key from the map. If the key doesn't exist, it returns ErrKeyNotExist
+func (g *GenericMap[K, V]) Delete(key *K) error {
+	return g.m.Delete(unsafe.Pointer(key))
+}
+
+// BatchDelete deletes a batch of keys from the map
+func (g *GenericMap[K, V]) BatchDelete(keys []K) (int, error) {
+	return g.m.BatchDelete(keys, nil)
+}
+
+// GenericMapIterator is an interface for iterating over a GenericMap
+type GenericMapIterator[K interface{}, V interface{}] interface {
+	// Next fills K and V with the next key/value pair in the map. It returns false if there are no more elements
+	Next(key *K, value *V) bool
+
+	// Err returns the last error that happened during iteration.
+	Err() error
+}
+
+func isPerCPU(t ebpf.MapType) bool {
+	switch t {
+	case ebpf.PerCPUHash, ebpf.PerCPUArray, ebpf.LRUCPUHash:
+		return true
+	}
+	return false
+}
+
+func (g *GenericMap[K, V]) isPerCPU() bool {
+	return isPerCPU(g.m.Type())
+}
+
+// Iterate returns an iterator for the map, which transparently chooses between batch and single item
+func (g *GenericMap[K, V]) Iterate() GenericMapIterator[K, V] {
+	return g.IterateWithOptions(IteratorOptions{BatchSize: defaultBatchSize})
+}
+
+func (g *GenericMap[K, V]) valueTypeCanUseUnsafePointer() bool {
+	// Simple test for now, but we probably will need to add more cases,
+	// as I am not 100% sure of the behavior of structs with maps
+	return !g.isPerCPU() // PerCPU maps use slices, so we need to pass them directly
+}
+
+// IterateWithOptions returns an iterator for the map, which transparently chooses between batch and single item
+// iterations. This version allows choosing options
+func (g *GenericMap[K, V]) IterateWithOptions(itops IteratorOptions) GenericMapIterator[K, V] {
+	if itops.BatchSize == 0 {
+		itops.BatchSize = defaultBatchSize // Default value for batch sizes. Possibly needs more testing to find an optimal default
+	}
+	if itops.BatchSize > int(g.m.MaxEntries()) {
+		itops.BatchSize = int(g.m.MaxEntries())
+	}
```

Comment:
we should have a warn here

Reply:
or even better - fail the operation

Reply:
Personally I wouldn't fail, right now the default is to iterate with batches (although [that could change](https://github.com/DataDog/datadog-agent/pull/21738#discussion_r1439238137)) and depending on the batch and map sizes we could be failing a lot of iterations by default.

Reply:
So a warning log (once) should be sufficient
My concern is that the naive user will try to increase the `BatchSize` field in `IteratrOptions` and once he passes a value larger than max he does not know, but also does not see any impact

Reply:
Or at least document the boundaries in the function's documentation (for BatchSize == 0 and BatchSize > MaxEntries)